### PR TITLE
Using cublasGemmBatchedEx/cublasGemmStridedBatchedEx for training

### PR DESCRIPTION
**Pull request by weixingzhang**
For some 1P training task, we found accuracy issue on V100. It turns out that the accumulation for matmul needs to be done in FP32 for training.

Here are the throughput of BERT-L on V100 16GB with Lamb. As expected, the perf is almost same. 
  | Batch Size | Seq Len | Throughput (ex/s)
-- | -- | -- | --
Before | 64 | 128 | 213.935
  | 10 | 512 | 41.9966
After | 64 | 128 | 213.228
  | 10 | 512 | 41.6287

### CODE REVIEW

File: onnxruntime/core/providers/cuda/cuda_common.h [hunk 1]
```diff
@@ -225,15 +225,20 @@ inline bool CalculateFdmStrides(gsl::span<fast_divmod> p, const std::vector<int6
 
 class CublasMathModeSetter {
  public:
-  CublasMathModeSetter(cublasHandle_t handle, cublasMath_t mode) : handle_(handle) {
-    cublasGetMathMode(handle, &mode_);
-    cublasSetMathMode(handle, mode);
+  CublasMathModeSetter(const cudaDeviceProp& prop,cublasHandle_t handle, cublasMath_t mode) : prop_(prop), handle_(handle) {
```
Comment (wschin) [inline onnxruntime/core/providers/cuda/cuda_common.h:228]: space

File: onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h [hunk 2]
```diff
@@ -18,49 +18,175 @@
...
+  return cublasHgemmBatched(handle,
+                            transa,
+                            transb,
+                            m, n, k,
+                            alpha,
+                            (const __half**)Aarray, lda,
+                            (const __half**)Barray, ldb,
+                            beta,
+                            (__half**)Carray, ldc,
+                            batch_count);
```
Comment (wschin) [inline onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h:188]: >batch_count); [](start = 28, length = 13)

nit: Maybe explicitly set CUDA_R_16F here to avoid confusion?

File: onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h [hunk 3]
```diff
@@ -18,49 +18,175 @@
...
+                            Aarray, lda,
+                            Barray, ldb,
+                            beta,
+                            Carray, ldc,
+                            batch_count);
+}
+inline cublasStatus_t cublasGemmBatchedHelper(cublasHandle_t handle,
+                                              cublasOperation_t transa,
+                                              cublasOperation_t transb,
+                                              int m, int n, int k,
```
Comment (wschin) [inline onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h:151]: >int m, int n, int k, [](start = 46, length = 20)

nit: some int parameters can be `const`.

File: onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h [hunk 4]
```diff
@@ -18,49 +18,175 @@
...
```
Comment (wschin) [inline onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h:129]: >int m, int n, int k, [](start = 46, length = 20)

Similar to other places, some parameters can be `const`.

File: onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h (hunk 3 above)
Comment (weixingzhang) [reply onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h:151]: What I learned is that const for built-in types in function parameters with pass-by-value is not necessary. See this link: https://abseil.io/tips/109

File: onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h (hunk 2 above)
Comment (weixingzhang) [reply onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h:188]: Maybe I misunderstand your comment. Were you saying to specify CUDA_R_16F when calling cublasHgemmBatched? It doesn't support to set data type.

File: onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h (hunk 3 above)
Comment (edgchen1) [reply onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h:151]: nice tip!
to quote from that link:
```
Do use top-level const on function parameters in definitions at your (or your team’s) discretion. You might follow the same rationale as you would for when to declare a function-local variable const.
```
i think const still has some usefulness here, similar to const local variables.

File: onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h [hunk 5]
```diff
@@ -18,49 +18,175 @@
...
+  float h_a = onnxruntime::math::halfToFloat(*reinterpret_cast<const uint16_t*>(alpha));
+  float h_b = onnxruntime::math::halfToFloat(*reinterpret_cast<const uint16_t*>(beta));
+
+  // accumulating in FP32
+  return cublasGemmBatchedEx(handle,
+                             transa,
+                             transb,
+                             m, n, k,
+                             &h_a,
+                             (const void**)Aarray, CUDA_R_16F, lda,
```
Comment (edgchen1) [inline onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h:170]: avoid C-style casts per https://google.github.io/styleguide/cppguide.html#Casting

File: onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h (hunk 3 above)
Comment (weixingzhang) [reply onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h:151]: I remember before when I use const to int as suggested here, the compiler would complain something like "it is not necessary to use const for built-in types.". Moreover, Nvidia doesn't use const for int in these cublas APIs either.

File: onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h (hunk 2 above)
Comment (wschin) [reply onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h:188]: `CUDA_R_32F` is a case. In another case, we can also explicitly provide that default argument (I wrongly through the default argument is `CUDA_R_16F`, sorry) for clarify.

File: onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h (hunk 3 above)
Comment (wschin) [reply onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h:151]: The major benefit of `const` is for readability to me (so I marked my comment as nit).

**Comment by SherlockNoMad** [comment]
Please run E2E test and "Pytorch Frontend E2E" test, I not sure if this the batchmatmul result is different enough to affect the expected test values.
